File: Python3.x/39-Combination Sum.py
# dfs
class Solution:
    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        self.tmp = []
        candidates.sort()
        self.dfs(candidates, [], target, 0)
        return self.tmp

    def dfs(self, candidates, tmp_list, target, n):
        if target == 0: self.tmp.append(tmp_list[:])
        if target < candidates[0]: return
        for i in candidates:
            if i > target: return
            if i < n: continue
            tmp_list.append(i)  # 栈
            self.dfs(candidates, tmp_list, target - i, i)
            tmp_list.pop()


'''
# 递归
class Solution:
    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        candidates.sort()
        output, length = [], len(candidates)
        def func(tmp, candidates, target, n):
            if sum(tmp) > target: return
            if sum(tmp) == target: return output.append(tmp)
            for i in range(n, length):
                func(tmp + [candidates[i]], candidates, target, i)

        func([], candidates, target, 0)
        return output
'''

File: C++/39-Combination_Sum.cpp
// Runtime: 8 ms, faster than 98.27% of C++ online submissions for Combination Sum.
// Memory Usage: 9.2 MB, less than 99.01% of C++ online submissions for Combination Sum.
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        vector<vector<int>> res;
        vector<int> curr;
        recursion(candidates, target, res, curr, 0);
        return res;
    }
private:
    void recursion(vector<int>& candidates, int target, vector<vector<int>>& res, 
                   vector<int>& curr, int idx) {
        if (!target) {
            res.push_back(curr);
            return;
        }
        for (int i = idx; i != candidates.size() && target >= candidates[i]; ++i) {
            curr.push_back(candidates[i]);
            recursion(candidates, target - candidates[i], res, curr, i);
            curr.pop_back();
        }
    }
};
